File: src/workspace.js
/**
 * Workspace v2 — output path resolution for plugin build scripts.
 *
 * Works at any directory depth by walking up to find markers
 * instead of assuming a fixed number of parent levels.
 *
 * Usage (from any build script inside workspace/):
 *   const path = require('path');
 *   let root = __dirname;
 *   while (!require('fs').existsSync(path.join(root, 'package.json'))) root = path.dirname(root);
 *   const { ensureOutputDir } = require(path.join(root, 'src/workspace'));
 *   const outputDir = ensureOutputDir(__dirname);
 */

const path = require('path');
const fs = require('fs');

/**
 * Walk up from startDir to find the repo root (directory containing package.json).
 * Skips node_modules directories.
 */
function findRepoRoot(startDir) {
  let dir = path.resolve(startDir);
  while (dir !== path.dirname(dir)) {
    if (
      fs.existsSync(path.join(dir, 'package.json')) &&
      !dir.includes('node_modules')
    ) {
      return dir;
    }
    dir = path.dirname(dir);
  }
  return null;
}

/**
 * Walk up from a build script directory to find the project root.
 *
 * The project root is the directory whose child named 'build' contains
 * the build script. Works at any depth:
 *   workspace/build/<deliverable>/                → workspace/
 *   workspace/<project>/build/<deliverable>/      → workspace/<project>/
 */
function resolveProjectRoot(buildDir) {
  let dir = path.resolve(buildDir);
  while (dir !== path.dirname(dir)) {
    const parent = path.dirname(dir);
    if (path.basename(parent) === 'build') {
      return path.dirname(parent);
    }
    dir = parent;
  }
  return path.resolve(buildDir);
}

/**
 * Walk up from startDir to find the workspace/ root directory.
 * Returns null if not inside a workspace tree.
 */
function resolveWorkspaceRoot(startDir) {
  let dir = path.resolve(startDir);
  while (dir !== path.dirname(dir)) {
    if (path.basename(dir) === 'workspace') return dir;
    dir = path.dirname(dir);
  }
  return null;
}

/**
 * Resolve the output directory for a build script.
 *
 * Precedence:
 *   1. options.outputDir (explicit — for external agents)
 *   2. OUTPUT_DIR env var
 *   3. .remotion-project marker → <projectRoot>/out/
 *   4. Convention → <projectRoot>/output/<deliverable>/
 */
function resolveOutputDir(buildDir, options) {
  if (options && options.outputDir) return path.resolve(options.outputDir);
  if (process.env.OUTPUT_DIR) return path.resolve(process.env.OUTPUT_DIR);

  const projectRoot = resolveProjectRoot(buildDir);
  const abs = path.resolve(buildDir);

  if (fs.existsSync(path.join(projectRoot, '.remotion-project'))) {
    return path.join(projectRoot, 'out');
  }

  const parent = path.dirname(abs);
  if (path.basename(parent) === 'build') {
    const deliverable = path.basename(abs);
    return path.join(projectRoot, 'output', deliverable);
  }

  return path.join(projectRoot, 'output');
}

/** Resolve output dir and create it if it doesn't exist. */
function ensureOutputDir(buildDir, options) {
  const dir = resolveOutputDir(buildDir, options);
  fs.mkdirSync(dir, { recursive: true });
  return dir;
}

/** Resolve full path to an output file. */
function resolveOutputPath(buildDir, filename, options) {
  return path.join(resolveOutputDir(buildDir, options), filename);
}

/** Resolve the intake directory for a project. */
function resolveIntakeDir(buildDir) {
  return path.join(resolveProjectRoot(buildDir), 'intake');
}

module.exports = {
  findRepoRoot,
  resolveProjectRoot,
  resolveWorkspaceRoot,
  resolveOutputDir,
  ensureOutputDir,
  resolveOutputPath,
  resolveIntakeDir,
};

File: src/workspace.py
"""
Workspace v2 — output path resolution for plugin build scripts.

Works at any directory depth by walking up to find markers
instead of assuming a fixed number of parent levels.

Usage (from any build script inside workspace/):
    from pathlib import Path
    root = Path(__file__).resolve()
    while not (root / 'package.json').exists():
        root = root.parent
    import sys; sys.path.insert(0, str(root / 'src'))
    from workspace import ensure_output_dir

    output_dir = ensure_output_dir(__file__)
"""

from __future__ import annotations

import os
from pathlib import Path


def find_repo_root(start_dir: str | Path) -> Path | None:
    """Walk up from start_dir to find the repo root (directory containing package.json).
    Skips node_modules directories.
    """
    p = Path(start_dir).resolve()
    while p != p.parent:
        if (p / "package.json").exists() and "node_modules" not in p.parts:
            return p
        p = p.parent
    return None


def resolve_project_root(build_path: str | Path) -> Path:
    """Walk up from a build script to find the project root.

    The project root is the directory whose child named 'build' contains
    the build script. Works at any depth:
        workspace/build/<deliverable>/            → workspace/
        workspace/<project>/build/<deliverable>/  → workspace/<project>/
    """
    p = Path(build_path).resolve()
    if p.is_file():
        p = p.parent

    d = p
    while d != d.parent:
        if d.name == "build":
            return d.parent
        d = d.parent
    return p


def resolve_workspace_root(start_dir: str | Path) -> Path | None:
    """Walk up from start_dir to find the workspace/ root directory.
    Returns None if not inside a workspace tree.
    """
    p = Path(start_dir).resolve()
    while p != p.parent:
        if p.name == "workspace":
            return p
        p = p.parent
    return None


def resolve_output_dir(
    build_path: str | Path, *, output_dir: str | Path | None = None
) -> Path:
    """Resolve the output directory for a build script.

    Precedence:
        1. output_dir kwarg (explicit — for external agents)
        2. OUTPUT_DIR env var
        3. Convention → <projectRoot>/output/<deliverable>/
    """
    if output_dir is not None:
        return Path(output_dir).resolve()
    env = os.environ.get("OUTPUT_DIR")
    if env:
        return Path(env).resolve()

    project_root = resolve_project_root(build_path)
    p = Path(build_path).resolve()
    if p.is_file():
        p = p.parent

    # Walk up to find the 'build' directory and extract the deliverable name
    d = p
    while d != d.parent:
        if d.parent.name == "build":
            return project_root / "output" / d.name
        if d.name == "build":
            break
        d = d.parent

    return project_root / "output"


def ensure_output_dir(
    build_path: str | Path, *, output_dir: str | Path | None = None
) -> Path:
    """Resolve output dir and create it if it doesn't exist."""
    d = resolve_output_dir(build_path, output_dir=output_dir)
    d.mkdir(parents=True, exist_ok=True)
    return d


def resolve_output_path(
    build_path: str | Path, filename: str, *, output_dir: str | Path | None = None
) -> Path:
    """Resolve full path to an output file."""
    return resolve_output_dir(build_path, output_dir=output_dir) / filename


def resolve_intake_dir(build_path: str | Path) -> Path:
    """Resolve the intake directory for a project."""
    return resolve_project_root(build_path) / "intake"
